### ISSUE TITLE
Incorrect type error in v2.7.1+ when using record index accessor to generate union of value types

### ISSUE THREAD

**Issue by WearyMonkey** [Microsoft/TypeScript#22116]
**TypeScript Version:**  
typescript@2.8.0-dev.20180222 (🔴 not working)
typescript@2.7.2 (🔴 not working)
typescript@2.7.1 (🔴 not working)
typescript@2.7.0-rc (💚  working)

**Search Terms:** 

union keyof interface inference generic

**Code**

```ts
interface Baa<Types> {
  foo: Types[keyof Types];
};

type A = Baa<{ a: string } & { b: string }>;
type B = Baa<{ a: string }>;

const _a: A = { foo: 'banana' };
const _b: B = _a;
```

```
{
  "compilerOptions": {
    "strictFunctionTypes": true
  }
}
```

**Expected behavior:**

No error, as typeof `A['foo']`  and `B['foo']` should both be `string`.

**Actual behavior:**

```
bad.ts(9,7): error TS2322: Type 'Baa<{ a: string; } & { b: string; }>' is not assignable to type 'Baa<{ a: string; }>'.
  Type '{ a: string; }' is not assignable to type '{ a: string; } & { b: string; }'.
    Type '{ a: string; }' is not assignable to type '{ b: string; }'.
      Property 'b' is missing in type '{ a: string; }'.
error Command failed with exit code 2.
```

**Playground Link:**  

https://www.typescriptlang.org/play/#src=interface%20Baa%3CTypes%3E%20%7B%0D%0A%09foo%3A%20Types%5Bkeyof%20Types%5D%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20a%3A%20Baa%3C%7B%20a%3A%20string%20%7D%20%26%20%7B%20b%3A%20string%20%7D%3E%20%3D%20%7B%20foo%3A%20'banana'%20%7D%3B%0D%0Aconst%20b%3A%20Baa%3C%7B%20a%3A%20string%20%7D%3E%20%3D%20a%3B%0D%0A

Notes:

Only occurs with interfaces and not types.

**Comment by ahejlsberg** [comment]
This one is a bit tricky. The reason it now errors for interfaces (but not type aliases) is that for generic classes and interfaces in `--strictFunctionTypes` mode we probe variance using exemplar types and then use that information to reduce type checking work later. In this case we determine that `Baa` is contra-variant for the `Types` parameter and thus we consider two references to `Baa` to be related only when their type arguments are related contra-variantly. That isn't the case in the example, so we issue an error. The variance probing is by nature a conservative estimate, and in certain cases it may be too conservative. This is one of those cases. Specifically, because the extra property in `A` is of an already existing type, `Types[keyof Types]` is actually the same for `A` and `B`, but we don't know that unless we do a full structural comparison--which we do for type aliases.

The reason it worked in `2.7.0-rc` was because we hadn't yet fixed a related error reporting bug. We would actually find the types incompatible but then fail to correctly report the error.

I think your best bet is to change the declaration to use a type alias as you will then get a full structural comparison every time.
